drivers/ethernet/phy/phy_lan8720a.c

```diff
@@ -0,0 +1,425 @@
+	if (speeds & LINK_FULL_100BASE_T) {
+		anar |= MII_ADVERTISE_100_FULL;
+	} else {
+		anar &= ~MII_ADVERTISE_100_FULL;
+	}
+	if (speeds & LINK_HALF_100BASE_T) {
+		anar |= MII_ADVERTISE_100_HALF;
+	} else {
+		anar &= ~MII_ADVERTISE_100_HALF;
+	}
+	if (speeds & LINK_FULL_10BASE_T) {
+		anar |= MII_ADVERTISE_10_FULL;
+	} else {
+		anar &= ~MII_ADVERTISE_10_FULL;
+	}
+	if (speeds & LINK_HALF_10BASE_T) {
+		anar |= MII_ADVERTISE_10_HALF;
+	} else {
+		anar &= ~MII_ADVERTISE_10_HALF;
+	}
+
+	/* Write capabilities to advertising register */
+	ret = phy_lan8720a_write(dev, MII_ANAR, anar);
+	if (ret < 0) {
+		LOG_ERR("Failed to write ANAR (%d)", ret);
+		goto unlock;
+	}
+
+	/* (re)do auto-negotiation */
+	ret = phy_lan8720a_auto_negotiate(dev);
+	if (ret < 0) {
+		LOG_ERR("Auto-negotiation error (%d)", ret);
+		goto unlock;
+	}
+
+	/* Get link status */
+	ret = phy_lan8720a_get_link(dev, &data->state);
+	if (ret < 0) {
+		LOG_ERR("Failed to get link status (%d)", ret);
+		goto unlock;
+	}
+
+	/* Log the results of the configuration */
+	LOG_INF("PHY %d is %s", config->addr, data->state.is_up ? "up" : "down");
+	LOG_INF("PHY (%d) Link speed %s Mb, %s duplex\n", config->addr,
+		(PHY_LINK_IS_SPEED_100M(data->state.speed) ? "100" : "10"),
+		PHY_LINK_IS_FULL_DUPLEX(data->state.speed) ? "full" : "half");
+
+unlock:
+	(void)k_mutex_unlock(&data->mutex);
+
+	return ret;
+}
+
+static int phy_lan8720a_link_cb_set(const struct device *dev, phy_callback_t cb, void *user_data)
+{
+	struct lan8720a_data *data = dev->data;
+	int ret;
+
+	data->cb = cb;
+	data->cb_data = user_data;
+
+	ret = phy_lan8720a_get_link(dev, &data->state);
+	if (ret < 0) {
+		return ret;
+	}
+
+	data->cb(dev, &data->state, data->cb_data);
+
+	return 0;
+}
+
+static void phy_lan8720a_isr(const struct device *port, struct gpio_callback *cb,
+			     gpio_port_pins_t pins)
+{
+	struct lan8720a_data *data = CONTAINER_OF(cb, struct lan8720a_data, irq_callback);
+
+	ARG_UNUSED(port);
+	ARG_UNUSED(pins);
+
+	k_work_submit(&data->irq_work);
+}
+
+static void phy_lan8720a_work_handler(struct k_work *work)
+{
+	struct lan8720a_data *data = CONTAINER_OF(work, struct lan8720a_data, irq_work);
+	struct phy_link_state new_state;
+	int ret;
+	uint32_t irq_value;
+
+	ret = phy_lan8720a_read(data->self, LAN8720A_INTERRUPT_SOURCE_REG, &irq_value);
+	if (ret < 0) {
+		LOG_ERR("Failed to read interrupt register (%d)", ret);
+		return;
+	}
+
+	ret = phy_lan8720a_get_link(data->self, &new_state);
+	if (ret < 0) {
+		return;
+	}
+
+	if (memcmp(&data->state, &new_state, sizeof(struct phy_link_state)) != 0) {
+		memcpy(&data->state, &new_state, sizeof(struct phy_link_state));
+
+		if (data->cb) {
+			data->cb(data->self, &data->state, data->cb_data);
+		}
+	}
+}
+
+static int phy_lan8720a_init(const struct device *dev)
+{
+	const struct lan8720a_config *cfg = dev->config;
+	struct lan8720a_data *data = dev->data;
+	int ret;
+
+	data->self = dev;
+
+	(void)k_mutex_init(&data->mutex);
```

Comment:
I would move that part below the gpios initializing:

<pre>
(void)k_mutex_init(&data->mutex);
k_work_init(&data->irq_work, phy_lan8720a_work_handler);

mdio_bus_enable(cfg->mdio_dev);
</pre>

Reply:
I've moved the mdio bus enable, the others should be kept at the start.